rfcs/0000-plan-dynanism.md

```diff
@@ -0,0 +1,235 @@
+In turn, I hope these upstream packages and ecosystems might even care about packaging and integration of the sort that we do.
+This would create a virtuous cycle where Nix is easier to use by more people, and Nixpkgs is easier to maintain as upstream packages better match our values.
+
+# Detailed design
+[design]: #detailed-design
+
+We can break this down nicely into steps.
+
+## Dynamic derivations
+
+*This is implemented in https://github.com/NixOS/nix/pull/4628.*
+
+1. Derivation outputs can be valid derivations:
+
+   1. Allow derivation outputs to be content addressed with the "text hashing" scheme.
+
+   2. Lift the restriction barring derivations output paths from ending in `.drv` if they are so content-addressed
+
+2. Extend the CLI to take advantage of such derivations:
+
+   We hopefully will soon allow CLI "installable" args in the form
+   ```
+   single-installable ::= <path> ! <output-name>
+   ```
+   where the first path is a derivation, and the second is the output we want to build.
+
+   We should generalize the grammar like so:
+   ```
+   single-installable ::= <single-installable> ! <output-name>
+                       |  <path>
+
+   multi-installable  ::= <single-installable>
+                       |  <single-installable> ! *
+   ```
+
+   Now that we have `path` vs `path!*`, we also don't need `---derivation` as a disambiguator, and so that should be removed along with all the complexity that goes with it.
+   (`toBuildables` in the the nix commands should always be pure functions and not consult the store.)
+
+3. Extend the scheduler and derivation dependencies similarly:
+
+  - Derivations can depended on the outputs of derivations that are themselves derivation outputs.
+    The schedule will substitute derivations to simplify dependencies as computed derivations are built, just like how floating content addressed derivations are realized.
+
+  - Missing derivations get their own full fledged goals so they can be built, not just fetched from substituters.
+
+4. Add a new `outputOf` primop:
+
+   `builtins.outputOf drv outputName` produces a placeholder string with the appropriate string context to access the output of that name produced by that derivation.
+   The placeholder string is quite analogous to that used for floating content addressed derivation outputs.
+   This isn't needed today when we statically know the derivation and thus it's outputs to make attributes for them, but it is needed for built derivations because we don't know either the final derivation or its output names in advanced.
+
+## Deferred import from derivation.
+
+Create a new primop `assumeDerivation` that takes a single expression.
+It's semantics are as follows:
+
+ - If the underlying expression evaluates (shallowly) without needing to build derivations (for `import`, `readFile`, etc.), and is a derivation, simply return that.
+   ```
+   e ⇓ e'
+   e' is derivation
+   ------
+   builtins.assumeDerivation e ⇓ e'
+   ```
```

Comment:
Could you add a section explaining this syntax? I assume it's familiar for PLT folks, but it's very cryptic to me :)

Reply:
Those are inference rules expressed in operational semantics.

You can read that as `if e evaluates to e' and e' is a derivation, then builtins.assumeDerivation e evaluates to e'`.

I agree with asymmetric, we're not a PLT-focused symposium, prose is fine here.

Reply:
I think a more accessible presentation would be better. This is pseudo-formal, because (AFAIK) the `⇓` and `is derivation` judgements aren't actually defined anywhere.

Reply:
Agreed. Unfortunately not everyone has that background.

Reply:
A "translation" would suffice. Those compaxt bits of formalisms are great, once one has learned to parse them.

Reply:
Well.....I'll put it this way:

1. I think CA derivations' various notions of store coherence (by far the subtlest part), GUIX-style "grafting", mixing native- and cross-compiled packages, and probably other things I can't anticipate are going to repeatedly keep us thinking about these sorts of inductive relations. Some sort of sequent-ish syntax, or prolog's horn clause syntax, or many other such things are far and away the clearest and most concise way to convey this information *once* the reader is familiar --- it's supposed to be nice and "pictoral" not inscrutable ivory tower. I think it be best if we start moving in that direction.

2. Even if I right some prose now, It's going to a pain to update. VS with this sort of syntax it can be easier to just stare at and have all the important bits come out, so it's easier to revise.

So I am fine writing some sort of prose explanation, but I rather do it once the RFC is further along.

----

P.S. whether you love our hate this stuff, https://www.youtube.com/watch?v=dCuZkaaou0Q is a fun talk.

Reply:
Since this is gone we can remove this.